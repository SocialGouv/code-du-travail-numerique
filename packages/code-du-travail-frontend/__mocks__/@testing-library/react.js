/* eslint-disable import/export */
import { theme } from "@socialgouv/cdtn-ui";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";
import React from "react";
import { ThemeProvider } from "styled-components";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme.colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
