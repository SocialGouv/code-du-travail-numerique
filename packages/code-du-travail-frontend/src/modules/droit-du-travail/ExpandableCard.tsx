"use client";

import React, { useState, useRef } from "react";
import { fr } from "@codegouvfr/react-dsfr";
import { css } from "@styled-system/css";
import Image from "next/image";

type ExpandableCardProps = {
  title: string;
  children: React.ReactNode;
  iconSrc: string;
  id?: string;
  backgroundColor?: string;
  showBottomTab?: boolean;
};

const ExpandableCard = ({
  title,
  children,
  iconSrc,
  id,
  backgroundColor,
  showBottomTab = false,
}: ExpandableCardProps) => {
  const [isExpanded, setIsExpanded] = useState(false);
  const contentRef = useRef<HTMLDivElement>(null);

  const toggleExpand = () => {
    setIsExpanded(!isExpanded);
  };

  return (
    <div className={cardContainer(backgroundColor, showBottomTab)} id={id}>
      <button
        onClick={toggleExpand}
        className={cardButton}
        aria-expanded={isExpanded}
        aria-controls={`content-${title.replace(/\s+/g, "-").toLowerCase()}`}
      >
        <div className={cardHeader}>
          <div className={iconContainer}>
            <Image
              src={iconSrc}
              alt=""
              width={52}
              height={52}
              aria-hidden="true"
            />
          </div>
          <h3 className={`${fr.cx("fr-mb-0", "fr-h4")} ${titleStyle}`}>
            {title}
          </h3>
          <div className={arrowContainer}>
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="24"
              height="24"
              viewBox="0 0 24 24"
              fill="none"
              stroke="currentColor"
              strokeWidth="2"
              strokeLinecap="round"
              strokeLinejoin="round"
              className={isExpanded ? expandedArrow : collapsedArrow}
              aria-hidden="true"
            >
              <polyline points="6 9 12 15 18 9"></polyline>
            </svg>
          </div>
        </div>
      </button>
      {isExpanded && (
        <div
          className={cardContent}
          id={`content-${title.replace(/\s+/g, "-").toLowerCase()}`}
          ref={contentRef}
        >
          {children}
        </div>
      )}
      {showBottomTab && (
        <div className={bottomTabContainer}>
          <div className={bottomTab(backgroundColor)}></div>
        </div>
      )}
    </div>
  );
};

export default ExpandableCard;

const cardContainer = (backgroundColor?: string, showBottomTab?: boolean) =>
  css({
    background: backgroundColor || "var(--background-alt-blue-cumulus)",
    overflow: showBottomTab ? "visible" : "hidden",
    height: "fit-content",
    transition: "all 0.3s ease",
    position: "relative",
    boxShadow: "0 2px 8px rgba(0, 0, 0, 0.1)",
  });

const titleStyle = css({
  color: "var(--text-action-high-blue-france)!",
});

const cardButton = css({
  width: "100%",
  height: "100%",
  background: "none",
  border: "none",
  padding: 0,
  cursor: "pointer",
  textAlign: "left",
  "&:hover": {
    background: "none!",
  },
});

const cardHeader = css({
  display: "flex",
  alignItems: "center",
  paddingX: "16px",
  position: "relative",
  height: "100%",
  paddingY: "40px",
});

const iconContainer = css({
  marginRight: "16px",
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
});

const arrowContainer = css({
  marginLeft: "auto",
});

const collapsedArrow = css({
  transition: "transform 0.3s ease",
});

const expandedArrow = css({
  transition: "transform 0.3s ease",
  transform: "rotate(180deg)",
});

const cardContent = css({
  padding: "0 16px 16px 16px",
});

const bottomTabContainer = css({
  position: "absolute",
  bottom: "-15px",
  left: "50%",
  transform: "translateX(-50%)",
  width: "100px", // Augmenté de 60px à 100px
  height: "30px",
  overflow: "visible",
});

const bottomTab = (backgroundColor?: string) =>
  css({
    width: "100%",
    height: "100%",
    background: backgroundColor || "var(--background-alt-blue-cumulus)",
    clipPath: "polygon(0% 0%, 100% 0%, 50% 100%)",
    filter: "drop-shadow(0 2px 8px rgba(0, 0, 0, 0.1))",
  });
