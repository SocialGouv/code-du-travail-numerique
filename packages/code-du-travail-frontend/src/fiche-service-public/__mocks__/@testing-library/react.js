import React from "react";
import PropTypes from "prop-types";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { theme } from "@socialgouv/cdtn-ui";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme.colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
