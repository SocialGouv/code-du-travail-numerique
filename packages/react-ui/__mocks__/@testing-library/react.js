import { render } from "@testing-library/react";
import PropTypes from "prop-types";
import React from "react";
import { ThemeProvider } from "styled-components";

import { colors } from "../../src/theme";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
