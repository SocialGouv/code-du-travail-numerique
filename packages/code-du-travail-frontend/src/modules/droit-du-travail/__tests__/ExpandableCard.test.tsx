import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import ExpandableCard from "../ExpandableCard";

// Mock Next.js Image component
jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line jsx-a11y/alt-text, @next/next/no-img-element
    return <img {...props} />;
  },
}));

describe("<ExpandableCard />", () => {
  const defaultProps = {
    title: "Test Title",
    iconSrc: "/test-icon.svg",
    children: <p>Test content</p>,
  };

  it("renders correctly with default props", () => {
    const { container } = render(<ExpandableCard {...defaultProps} />);
    expect(screen.getByText("Test Title")).toBeInTheDocument();
    expect(container.querySelector("img")).toHaveAttribute(
      "src",
      "/test-icon.svg"
    );
    expect(container).toMatchSnapshot();
  });

  it("does not show content when initially rendered", () => {
    render(<ExpandableCard {...defaultProps} />);
    expect(screen.queryByText("Test content")).not.toBeInTheDocument();
  });

  it("shows content when expanded", () => {
    render(<ExpandableCard {...defaultProps} />);

    // Click the button to expand
    fireEvent.click(screen.getByRole("button"));

    // Content should now be visible
    expect(screen.getByText("Test content")).toBeInTheDocument();
  });

  it("hides content when collapsed after being expanded", () => {
    render(<ExpandableCard {...defaultProps} />);

    // Click to expand
    fireEvent.click(screen.getByRole("button"));
    expect(screen.getByText("Test content")).toBeInTheDocument();

    // Click again to collapse
    fireEvent.click(screen.getByRole("button"));
    expect(screen.queryByText("Test content")).not.toBeInTheDocument();
  });

  it("renders with custom background color", () => {
    const { container } = render(
      <ExpandableCard
        {...defaultProps}
        backgroundColor="var(--background-alt-grey)"
      />
    );

    // The first div should have the background color
    const cardContainer = container.firstChild as HTMLElement;
    expect(cardContainer).toHaveStyle("background: var(--background-alt-grey)");
  });

  it("renders with bottom tab when showBottomTab is true", () => {
    const { container } = render(
      <ExpandableCard {...defaultProps} showBottomTab={true} />
    );

    // Instead of checking for the CSS class, we'll check if the component renders
    // an additional div when showBottomTab is true
    // The bottomTab is rendered as the last child of the container
    const expandedContent = screen.queryByText("Test content");
    expect(expandedContent).not.toBeInTheDocument(); // Content is not visible initially

    // Click to expand
    fireEvent.click(screen.getByRole("button"));

    // Now check the structure - there should be a div after the content div
    // when showBottomTab is true
    const cardContainer = container.firstChild as HTMLElement;
    expect(cardContainer.childNodes.length).toBeGreaterThan(2); // Button, content, and bottom tab
  });

  it("does not render bottom tab when showBottomTab is false", () => {
    const { container } = render(
      <ExpandableCard {...defaultProps} showBottomTab={false} />
    );

    // Check that the bottom tab div doesn't exist
    const bottomTab = container.querySelector('[class*="bottomTab"]');
    expect(bottomTab).not.toBeInTheDocument();
  });

  it("uses provided id attribute", () => {
    const { container } = render(
      <ExpandableCard {...defaultProps} id="test-id" />
    );

    const cardContainer = container.firstChild as HTMLElement;
    expect(cardContainer).toHaveAttribute("id", "test-id");
  });
});
