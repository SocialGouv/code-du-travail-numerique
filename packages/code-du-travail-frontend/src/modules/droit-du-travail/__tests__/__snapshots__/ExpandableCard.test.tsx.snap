// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpandableCard /> renders correctly with default props 1`] = `
<div>
  <div
    class="bg_var(--background-alt-blue-cumulus) bdr_8px ov_hidden h_fit-content trs_all_0.3s_ease pos_relative"
  >
    <button
      aria-controls="content-test-title"
      aria-expanded="false"
      class="w_100% bg_none bd_none p_0 cursor_pointer ta_left"
    >
      <div
        class="d_flex ai_center p_16px pos_relative"
      >
        <div
          class="mr_16px d_flex ai_center jc_center"
        >
          <img
            alt=""
            aria-hidden="true"
            height="40"
            src="/test-icon.svg"
            width="40"
          />
        </div>
        <h3
          class="fr-mb-0 fr-text--xl c_var(--text-action-high-blue-france)!"
        >
          Test Title
        </h3>
        <div
          class="ml_auto"
        >
          <svg
            aria-hidden="true"
            class="trs_transform_0.3s_ease"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </div>
      </div>
    </button>
  </div>
</div>
`;
